File: src/AnalogGameEngine/SimpleGUI/GameWindow.fs
namespace AnalogGameEngine.SimpleGUI

open OpenTK.Graphics
open OpenTK.Graphics.OpenGL4
open OpenTK.Mathematics
open OpenTK.Windowing
open OpenTK.Windowing.GraphicsLibraryFramework

open System

open AnalogGameEngine.SimpleGUI.Helper

type Textures = { cardBack: Texture; table: Texture }

type VertexData =
    {
        card: VertexDataObject
        cardBack: VertexDataObject
        stack: VertexDataObject
        table: VertexDataObject
    }

type GameWindow(gameWindowSettings, nativeWindowSettings) =
    inherit Desktop.GameWindow(gameWindowSettings, nativeWindowSettings)

    let mutable time = 0.
    let mutable keyNr1 = false

    let mutable shader: Shader option = None
    let mutable textures: Textures option = None
    let mutable vertexData: VertexData option = None

    let configureGL () =
        GL.ClearColor(0.2f, 0.2f, 0.2f, 1.f)
        GL.Enable(EnableCap.DepthTest)
        GL.Enable(EnableCap.CullFace)
        GL.Enable(EnableCap.Multisample)

    let initializeShader () =
        // TODO: workaround to make shaders work for now
        let shader =
            Shader.create "AnalogGameEngine.SimpleGUI/shader.vert" "AnalogGameEngine.SimpleGUI/shader.frag"

        Shader.useIt shader
        Shader.setInt shader "texture0" 0
        shader

    let initializeTextures (_: Shader) =
        {
            cardBack = Texture.create "assets/textures/Playingcards/CardBack_Red.jpg"
            table = Texture.create "assets/textures/Wood_006_COLOR.jpg"
        }

    let initializeVertexData (textures: Textures) =
        {
            table =
                // 3 position + 2 tex coords
                let topRight = [| 3.; 0.; -3.; 1.; 1. |]
                let topLeft = [| -3.; 0.; -3.; 0.; 1. |]
                let bottomLeft = [| -3.; 0.; 3.; 0.; 0. |]
                let bottomRight = [| 3.; 0.; 3.; 1.; 0. |]

                {
                    vertices =
                        [|
                            yield! topRight
                            yield! topLeft
                            yield! bottomLeft
                            yield! bottomRight
                        |]
                    config = [| 3; 2 |]
                    drawType = PrimitiveType.TriangleStrip
                    texture = textures.table
                    bufferObjectIndices = None
                }
                |> VertexDataObject.create
            cardBack =
                // 3 position + 2 tex coords
                let topRight = [| -0.315; 0.44; 0.; 1.; 1. |]
                let topLeft = [| 0.315; 0.44; 0.; 0.; 1. |]
                let bottomLeft = [| -0.315; -0.44; 0.; 1.; 0. |]
                let bottomRight = [| 0.315; -0.44; 0.; 0.; 0. |]

                {
                    vertices =
                        [|
                            yield! topRight
                            yield! topLeft
                            yield! bottomLeft
                            yield! bottomRight
                        |]
                    config = [| 3; 2 |]
                    drawType = PrimitiveType.TriangleStrip
                    texture = textures.cardBack
                    bufferObjectIndices = None
                }
                |> VertexDataObject.create
            stack =
                // 3 position + 2 tex coords
                let frontBottomRight = [| 0.315; 0.003; 0.44; 1.; 0. |]
                let frontBottomLeft = [| -0.315; 0.003; 0.44; 0.; 0. |]
                let frontTopRight = [| 0.315; 0.003; -0.44; 1.; 1. |]
                let frontTopLeft = [| -0.315; 0.003; -0.44; 0.; 1. |]
                let backBottomRight = [| 0.315; 0.; 0.44; 1.; 0. |]
                let backBottomLeft = [| -0.315; 0.; 0.44; 0.; 0. |]
                let backTopRight = [| 0.315; 0.; -0.44; 1.; 1. |]
                let backTopLeft = [| -0.315; 0.; -0.44; 0.; 1. |]

                {
                    vertices =
                        [|
                            yield! frontBottomRight
                            yield! frontBottomLeft
                            yield! frontTopRight
                            yield! frontTopLeft
                            yield! backBottomRight
                            yield! backBottomLeft
                            yield! backTopRight
                            yield! backTopLeft
                        |]
                    config = [| 3; 2 |]
                    drawType = PrimitiveType.TriangleStrip
                    texture = textures.cardBack
                    bufferObjectIndices =
                        Some [| 3u
                                1u
                                2u
                                0u
                                4u
                                1u
                                5u
                                3u
                                7u
                                2u
                                6u
                                4u |]
                }
                |> VertexDataObject.create
            card =
                // 3 position + 2 tex coords
                let topRight = [| 0.315; 0.44; 0.; 1.; 1. |]
                let topLeft = [| -0.315; 0.44; 0.; 0.; 1. |]
                let bottomRight = [| 0.315; -0.44; 0.; 1.; 0. |]
                let bottomLeft = [| 0.315; -0.44; 0.; 0.; 0. |]

                {
                    vertices =
                        [|
                            yield! topRight
                            yield! topLeft
                            yield! bottomLeft
                            yield! bottomRight
                        |]
                    config = [| 3; 2 |]
                    drawType = PrimitiveType.TriangleStrip
                    texture = textures.cardBack
                    bufferObjectIndices = None
                }
                |> VertexDataObject.create
        }

    let draw () =
        // Draw table
        let mutable model = Matrix4.Identity
        Shader.setMatrix4 shader.Value "model" model

        VertexDataObject.draw vertexData.Value.table None

    override _.OnLoad() =
        configureGL ()
        let newShader = initializeShader ()
        shader <- Some newShader
        let newTextures = initializeTextures newShader
        textures <- Some newTextures
        vertexData <- initializeVertexData newTextures |> Some

        base.OnLoad()

    override this.OnRenderFrame e =
        time <- time + (float e.Time)

        GL.Clear(
            ClearBufferMask.ColorBufferBit
            ||| ClearBufferMask.DepthBufferBit
        )

        Shader.useIt shader.Value

        // TODO: remove test code
        // TODO: move matrices to own camera class
        let radius = 7.0
        let camX = single (Math.Sin(time / 2.) * radius)
        let camZ = single (Math.Cos(time / 2.) * radius)

        let view =
            Matrix4.LookAt(Vector3(camX, 2.5f, camZ), Vector3(0.f, 0.f, 0.f), Vector3(0.f, 1.f, 0.f))

        Shader.setMatrix4 shader.Value "view" view

        let mutable projection = Matrix4.Identity

        projection <-
            projection
            * Matrix4.CreatePerspectiveFieldOfView(
                MathHelper.DegreesToRadians(45) |> single,
                this.ClientRectangle.Size.X
                / this.ClientRectangle.Size.Y
                |> single,
                0.1f,
                100.f
            )

        Shader.setMatrix4 shader.Value "projection" projection

        draw ()

        base.SwapBuffers()

        base.OnRenderFrame e

    override this.OnUpdateFrame e =
        let mouse = base.MouseState
        let input = base.KeyboardState

        if input.IsKeyDown(Keys.Escape) then
            this.Close()

        if not keyNr1 && input.IsKeyDown(Keys.D1) then
            if GL.IsEnabled EnableCap.CullFace then
                GL.Disable EnableCap.CullFace
            else
                GL.Enable EnableCap.CullFace

            keyNr1 <- true

        if keyNr1 && input.IsKeyReleased(Keys.D1) then
            keyNr1 <- false

        base.OnUpdateFrame e

    override _.OnResize e =
        GL.Viewport(0, 0, base.Size.X, base.Size.Y)
        base.OnResize e

    // Now, for cleanup. This isn't technically necessary since C# will clean up
    // all resources automatically when the program closes, but it's very
    // important to know how anyway
    override _.OnUnload() =
        // Unbind all the resources by binding the targets to 0/null.
        GL.BindBuffer(BufferTarget.ArrayBuffer, 0)
        GL.BindVertexArray 0
        GL.UseProgram 0

        // Delete all the resources.
        VertexDataObject.unload vertexData.Value.card
        VertexDataObject.unload vertexData.Value.cardBack
        VertexDataObject.unload vertexData.Value.table

        (shader.Value :> IDisposable).Dispose()
        base.OnUnload()

module GameWindow =
    let create width height title =
        let gameWindowSettings = new Desktop.GameWindowSettings()
        let nativeWindowsSettings = new Desktop.NativeWindowSettings()
        nativeWindowsSettings.Size <- new Vector2i(width, height)
        nativeWindowsSettings.Title <- title
        new GameWindow(gameWindowSettings, nativeWindowsSettings)

// TODO: remove
module Program =
    let window = GameWindow.create 800 600 "Test"
    window.Run()

File: old/AnalogGameEngine.SimpleGUI/GameWindow.cs
using OpenTK;
using OpenTK.Graphics;
using OpenTK.Graphics.OpenGL4;
using OpenTK.Input;
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Diagnostics;
using System.Linq;

using AnalogGameEngine.Entities;

using AnalogGameEngine.SimpleGUI.Helper;
using AnalogGameEngine.SimpleGUI.Entities;

namespace AnalogGameEngine.SimpleGUI {
    public class GameWindow<T, U> : OpenTK.GameWindow where T : SimpleGuiCard<U> where U : SimpleGuiCardType {
        private readonly GameBase<T> game;

        /* 1 = 1dm in real world */
        private VertexDataObject card, cardBack, stack, table;

        private Shader shader;
        private Texture cardbackTexture, tableTexture;

        private float time = 0.0f;
        private float deltaTime = 0.0f;
        private float lastFrame = 0.0f;

        private bool keyNr1 = false;

        public GameWindow(GameBase<T> game, int width, int height, string title)
            : base(
                width,
                height,
                GraphicsMode.Default,
                title,
                GameWindowFlags.Default,
                DisplayDevice.Default,
                4, 5,
                GraphicsContextFlags.Default
            ) {
            this.game = game;
        }

        protected override void OnLoad(EventArgs e) {
            ConfigureGL();
            shader = InitializeShader();
            this.InitializeTextures(shader);
            this.InitializeVertexData();

            base.OnLoad(e);
        }

        protected static void ConfigureGL() {
            GL.ClearColor(0.2f, 0.2f, 0.2f, 1.0f);
            GL.Enable(EnableCap.DepthTest);
            GL.Enable(EnableCap.CullFace);
            GL.Enable(EnableCap.Multisample);
        }

        protected static Shader InitializeShader() {
            // TODO: workaround to make shaders work for now
            var shader = new Shader("AnalogGameEngine.SimpleGUI/shader.vert", "AnalogGameEngine.SimpleGUI/shader.frag");
            shader.Use();
            shader.SetInt("texture0", 0);
            return shader;
        }

        // TODO: make more functional and add static
        protected void InitializeTextures(Shader shader) {
            // TODO: workaround to make textures work for now
            cardbackTexture = new Texture("Assets/Textures/Playingcards/CardBack_Red.jpg");
            tableTexture = new Texture("Assets/Textures/Wood_006_COLOR.jpg");
        }

        // TODO: make more functional and add static
        protected void InitializeVertexData() {
            table =
                new VertexDataObjectBuilder(
                    new[] {
                        // positions   // tex coords
                         3f, 0f, -3f,  1f, 1f, // top right
                        -3f, 0f, -3f,  0f, 1f, // top left
                        -3f, 0f,  3f,  0f, 0f, // bottom left
                         3f, 0f,  3f,  1f, 0f, // bottom right
                    },
                    new[] { 3, 2 },
                    PrimitiveType.TriangleFan
                )
                .WithTexture(tableTexture)
                .Build();

            cardBack =
                new VertexDataObjectBuilder(
                    new[] {
                        // positions          // tex coords
                        -0.315f,  0.44f, 0f,  1f, 1f, // top right
                         0.315f,  0.44f, 0f,  0f, 1f, // top left
                        -0.315f, -0.44f, 0f,  1f, 0f, // bottom right
                         0.315f, -0.44f, 0f,  0f, 0f, // bottom left
                    },
                    new[] { 3, 2 },
                    PrimitiveType.TriangleStrip
                )
                .WithTexture(cardbackTexture)
                .Build();

            stack =
                new VertexDataObjectBuilder(
                    new[] {
                        // positions              // tex coords
                         0.315f, 0.003f,  0.44f,  1f, 0f, // front bottom right
                        -0.315f, 0.003f,  0.44f,  0f, 0f, // front bottom left
                         0.315f, 0.003f, -0.44f,  1f, 1f, // front top right
                        -0.315f, 0.003f, -0.44f,  0f, 1f, // front top left
                         0.315f, 0f,      0.44f,  1f, 0f, // back bottom right
                        -0.315f, 0f,      0.44f,  0f, 0f, // back bottom left
                         0.315f, 0f,     -0.44f,  1f, 1f, // back top right
                        -0.315f, 0f,     -0.44f,  0f, 1f, // back top left
                    },
                    new[] { 3, 2 },
                    PrimitiveType.TriangleStrip
                )
                .WithTexture(cardbackTexture)
                .WithElementBufferObject(
                    new uint[] { 3, 1, 2, 0, 4, 1, 5, 3, 7, 2, 6, 4 }
                )
                .Build();

            card =
                new VertexDataObjectBuilder(
                    new[] {
                        // positions          // tex coords
                         0.315f,  0.44f, 0f,  1f, 1f, // top right
                        -0.315f,  0.44f, 0f,  0f, 1f, // top left
                         0.315f, -0.44f, 0f,  1f, 0f, // bottom right
                        -0.315f, -0.44f, 0f,  0f, 0f, // bottom left
                    },
                    new[] { 3, 2 },
                    PrimitiveType.TriangleStrip
                )
                .WithTexture(cardbackTexture)
                .Build();
        }

        protected override void OnRenderFrame(FrameEventArgs e) {
            time += (float)e.Time;

            GL.Clear(ClearBufferMask.ColorBufferBit | ClearBufferMask.DepthBufferBit);
            shader.Use();

            // TODO: remove test code
            // TODO: move matrices to own camera class
            float radius = 7.0f;
            float camX = (float)Math.Sin(time / 2) * radius;
            float camZ = (float)Math.Cos(time / 2) * radius;

            Matrix4 view = Matrix4.LookAt(new Vector3(camX, 2.5f, camZ), new Vector3(0f, 0f, 0f), new Vector3(0f, 1f, 0f));
            shader.SetMatrix4("view", view);

            Matrix4 projection = Matrix4.Identity;
            projection *= Matrix4.CreatePerspectiveFieldOfView(MathHelper.DegreesToRadians(45f), this.ClientRectangle.Width / this.ClientRectangle.Height, 0.1f, 100f);
            shader.SetMatrix4("projection", projection);

            this.Draw();

            Context.SwapBuffers();

            base.OnRenderFrame(e);
        }

        protected void Draw() {
            Matrix4 model;

            // Draw table
            model = Matrix4.Identity;
            shader.SetMatrix4("model", model);

            this.table.Draw();

            // Draw game stacks
            int stackAmount = game.Stacks.Count;
            for (int i = 0; i < stackAmount; i++) {
                var stack = game.Stacks.ElementAt(i).Value;
                if (stack.Count > 0) {
                    model = Matrix4.Identity;
                    model *= Matrix4.CreateScale(1f, stack.Count, 1f);
                    model *= Matrix4.CreateTranslation(0f, 0f, 0.8f);
                    model *= Matrix4.CreateRotationY(MathHelper.DegreesToRadians(i * (360f / stackAmount)));
                    shader.SetMatrix4("model", model);

                    this.stack.Draw();
                }
            }

            // Draw handcards
            float spacingX = 0.1f;
            float spacingY = 0.02f;
            float spacingZ = 0.001f;

            Matrix4 handBase = Matrix4.Identity;
            handBase *= Matrix4.CreateRotationX(MathHelper.DegreesToRadians(-20));
            handBase *= Matrix4.CreateTranslation(0f, 1.5f, 3f);

            int playerAmount = game.Players.Count;
            for (int i = 0; i < playerAmount; i++) {
                Matrix4 hand = handBase;
                hand *= Matrix4.CreateRotationY(MathHelper.DegreesToRadians(i * (360 / playerAmount)));

                // TODO: Add support for more private sets
                Set<T> handSet = game.Players[i].Sets.First().Value;
                int cardAmount = handSet.Cards.Count;
                for (int j = 0; j < cardAmount; j++) {
                    T card = handSet.Cards.ElementAt(j);
                    float spacingIndex = (j - cardAmount / 2f);

                    model = Matrix4.Identity;
                    model *= Matrix4.CreateTranslation(spacingIndex * spacingX, spacingIndex * spacingY, spacingIndex * spacingZ);
                    model *= hand;
                    shader.SetMatrix4("model", model);

                    this.card.Draw(card.Type.Texture);
                    this.cardBack.Draw();
                }
            }
        }

        protected override void OnUpdateFrame(FrameEventArgs e) {
            MouseState mouse = Mouse.GetState();
            KeyboardState input = Keyboard.GetState();

            if (input.IsKeyDown(Key.Escape)) {
                Exit();
            }
            if (!keyNr1 && input.IsKeyDown(Key.Number1)) {
                if (GL.IsEnabled(EnableCap.CullFace)) {
                    GL.Disable(EnableCap.CullFace);
                }
                else {
                    GL.Enable(EnableCap.CullFace);
                }
                keyNr1 = true;
            }
            if (keyNr1 && input.IsKeyUp(Key.Number1)) { keyNr1 = false; }

            base.OnUpdateFrame(e);
        }

        protected override void OnResize(EventArgs e) {
            GL.Viewport(0, 0, Width, Height);
            base.OnResize(e);
        }

        // Now, for cleanup. This isn't technically necessary since C# will clean up all resources automatically when the program closes, but it's very
        // important to know how anyway.
        protected override void OnUnload(EventArgs e) {
            // Unbind all the resources by binding the targets to 0/null.
            GL.BindBuffer(BufferTarget.ArrayBuffer, 0);
            GL.BindVertexArray(0);
            GL.UseProgram(0);

            // Delete all the resources.
            card.Unload();
            cardBack.Unload();
            table.Unload();

            shader.Dispose();
            base.OnUnload(e);
        }
    }
}
